src/Native/System.Native/pal_exec.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "../config.h"
+#include "pal_exec.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <unistd.h>
+
+enum
+{
+    READ_END_OF_PIPE = 0,
+    WRITE_END_OF_PIPE = 1,
+};
+    
+inline static void CloseIfOpen(int fd)
+{
+    if (fd >= 0)
+        close(fd);
+}
+
+extern "C"
+int32_t ForkAndExecProcess(
+    const char* filename,     // filename argument to execve
+    char* const argv[],       // argv argument to execve
+    char* const envp[],       // envp argument to execve
+    const char* cwd,          // path passed to chdir in child process
+    int32_t redirectStdin,    // whether to redirect standard input from the parent
+    int32_t redirectStdout,   // whether to redirect standard output to the parent
+    int32_t redirectStderr,   // whether to redirect standard error to the parent
+    int32_t* childPid,        // [out] the child process' id
+    int32_t* stdinFd,         // [out] if redirectStdin, the parent's fd for the child's stdin
+    int32_t* stdoutFd,        // [out] if redirectStdout, the parent's fd for the child's stdout
+    int32_t* stderrFd)        // [out] if redirectStderr, the parent's fd for the child's stderr
+{
+    int success = true;
+    int stdinFds[2] = { -1, -1 }, stdoutFds[2] = { -1, -1 }, stderrFds[2] = { -1, -1 };
+    int processId = -1;
+
+    // Validate arguments
+    if (nullptr == filename || nullptr == argv || nullptr == envp ||
+        nullptr == stdinFd || nullptr == stdoutFd || nullptr == stderrFd ||
+        nullptr == childPid)
+    {
+        assert(!"null argument.");
+        errno = EINVAL;
+        success = false;
+        goto done;
+    }
+    if ((redirectStdin  & ~1) != 0 ||
+        (redirectStdout & ~1) != 0 ||
+        (redirectStderr & ~1) != 0)
+    {
+        assert(!"Boolean redirect* inputs must be 0 or 1. ");
+        errno = EINVAL;
+        success = false;
+        goto done;
+    }
+
+    // Open pipes for any requests to redirect stdin/stdout/stderr
+    if ((redirectStdin  && pipe(stdinFds)  != 0) ||
+        (redirectStdout && pipe(stdoutFds) != 0) ||
+        (redirectStderr && pipe(stderrFds) != 0))
+    {
+        assert(!"pipe() failed.");
+        success = false;
+        goto done;
+    }
+
+    // Fork the child process
+    if ((processId = fork()) == -1)
+    {
+        assert(!"fork() failed.");
+        success = false;
+        goto done;
+    }
+
+    /* From the time the child process (processId == 0) begins running from fork to when
+     * it reaches execve, the child process must not touch anything in the PAL.  Doing so
+     * is not safe. The parent process (processId >= 0) may continue to use the PAL.
+     */
+
+    if (processId == 0) // processId == 0 if this is child process
+    {
+        // Close the parent end of any open pipes
```

Comment:
I realize I wrote this comment in the first place :), but we should probably clarify it slightly to say that we're closing the child's copy of the parent's end of any open pipes, not that we're actually closing the parent's end.